### PR TITLE
Client refactor

### PR DESCRIPTION
**Pull request by 3lnc**
# Fixes # 

### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [ ] I have added in line documentation to the code I modified

### Short description of what this PR does:
Better signatures for client; minor refactoring

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/python-http-client?pullRequest=88) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/python-http-client?pullRequest=88) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/python-http-client?pullRequest=88) it.</sub>

**Comment by thinkingserious** [comment]
Thank you @3lnc! I've added this to our backlog for a code review.

File: python_http_client/client.py [hunk 1]
```diff
@@ -62,6 +63,9 @@ def to_dict(self):
 class Client(object):
     """Quickly and easily access any REST or REST-like API."""
 
+    # These are the supported HTTP verbs
+    methods = set(('delete', 'get', 'patch', 'post', 'put'))
```
Comment (42B) [inline python_http_client/client.py]: `set` literal instead?

```python
methods = {'delete', 'get', 'patch', 'post', 'put'}
```

File: tests/test_unit.py [hunk 2]
```diff
@@ -84,7 +90,7 @@ def test__init__(self):
         self.assertEqual(default_client.host, self.host)
         self.assertEqual(default_client.request_headers, {})
         self.assertIs(default_client.timeout, None)
-        methods = ['delete', 'get', 'patch', 'post', 'put']
+        methods = set(('delete', 'get', 'patch', 'post', 'put'))
```
Comment (42B) [inline tests/test_unit.py]: `set` literal?

```python
methods = {'delete', 'get', 'patch', 'post', 'put'}
```

File: tests/test_unit.py [hunk 3]
```diff
@@ -97,7 +103,7 @@ def test__init__(self):
                         timeout=10)
         self.assertEqual(client.host, self.host)
         self.assertEqual(client.request_headers, request_headers)
-        methods = ['delete', 'get', 'patch', 'post', 'put']
+        methods = set(('delete', 'get', 'patch', 'post', 'put'))
```
Comment (42B) [inline tests/test_unit.py]: `set` literal?

```python
methods = {'delete', 'get', 'patch', 'post', 'put'}
```

File: tests/test_unit.py [hunk 4]
```diff
@@ -123,6 +129,19 @@ def test__build_url(self):
         built_url = self.client._build_url(query_params)
         self.assertEqual(built_url, url)
 
+    @mock.patch('python_http_client.client.Client._make_request')
+    def test__urllib_headers(self, maker):
+        self.client._update_headers({'X-test': 'Test'})
+        self.client.get()
+        request = maker.call_args[0][1]
+        self.assertTrue('X-test' in request.headers)
```
Comment (42B) [inline tests/test_unit.py]: `self.assertIn('X-test', request.headers)` ?

* https://docs.python.org/3/library/unittest.html?highlight=assertin#unittest.TestCase.assertIn

File: python_http_client/client.py (hunk 1 above)
Comment (3lnc) [reply python_http_client/client.py]: At time this PR was created, http client claimed support for 2.6

File: tests/test_unit.py (hunk 4 above)
Comment (3lnc) [reply tests/test_unit.py]: Available since 3.1, client provides support for 2.7 (2.6 previously)

**Comment by misterdorm** [comment]
@3lnc Please address conflicts and code coverage failures.  Thanks!

File: tests/test_unit.py (hunk 4 above)
Comment (42B) [reply tests/test_unit.py]: @3lnc 

> New in version 2.7.

* https://docs.python.org/2/library/unittest.html?highlight=assertin#unittest.TestCase.assertIn

File: python_http_client/client.py (hunk 1 above)
Comment (42B) [reply python_http_client/client.py]: @3lnc 

Since that doesn't seem to be the case now... set literal?

Comment (thinkingserious) [reply python_http_client/client.py]: These are the supported versions going forward: https://github.com/sendgrid/sendgrid-python/blob/v4/.travis.yml#L4

**Comment by thinkingserious** [comment]
Hi @3lnc,

When you get a moment to make the updates I'd love to merge this before we release v4 of the sendgrid-python library which will depend on this library.

Thanks!

With Best Regards,

Elmer

File: python_http_client/client.py [hunk 5]
```diff
@@ -211,44 +213,51 @@ def get_version(*args, **kwargs):
...
                 opener = urllib.build_opener()
-                request = urllib.Request(self._build_url(params), data=data)
-                if self.request_headers:
-                    for key, value in self.request_headers.items():
-                        request.add_header(key, value)
-                if data and ('Content-Type' not in self.request_headers):
-                    request.add_header('Content-Type', 'application/json')
+                request = urllib.Request(
+                    self._build_url(query_params),
+                    headers=self.request_headers,
```
Comment (stewartjarod) [inline python_http_client/client.py]: This fixes an issue I am having when adding headers... `request.add_header(key, value)` capitalizes the header which some APIs don't seem to want. :-\

**Comment by thinkingserious** [comment]
Hello @3lnc,<br /><br/>Thanks again for the PR!<br />                                              <br />We appreciate your contribution and look forward to                                              continued collaboration. Thanks!                                             <br /><br />Team SendGrid DX

**Pull request merged**